File: Competitive Programming/codeforces/training/D_General_Weighted_Max_Matching.cpp
// Copyright (c) 2023, Le Duc Phuc Long

/*If you don't think twice, you have to code twice.*/

// Include
#include <bits/stdc++.h>  // NOLINT
// #include <conio.h>

// Namespace
using namespace std;  // NOLINT

// Define
#define nln '\n'

// Typedef and Struct
typedef int64_t lo;

// Constant

// Declare global variables.
vector<pair<lo, lo>> dir4 = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};
// End of global variable declaration.

void precompute() {
}

void solve() {
    lo n;   cin >>n;
    vector<vector<lo>> drr(n, vector<lo>(n, 0));
    for (lo i = 0; i < n-1; ++i) {
        for (lo j = i+1; j < n; ++j)
            cin >> drr[i][j];
    }

    vector<lo> dp((1<<n), 0);
    for (lo i = 0; i < (1 << n); ++i) {
        lo j;
        for (j = 0; j < n; ++j)
            if (((i>>j) & 1) == 0)
                break;
        if (j == n) continue;
        for (lo k = j; k < n; ++k) {
                if (((i>>k) & 1) == 0) {
                    lo next_status = i | (1<<j) | (1<<k);
                    dp[next_status] = max(dp[next_status], dp[i]+drr[j][k]);
                }
            }
    }
    cout << dp.back();
}

int main(int argc, char* argv[]) {
    // freopen("in.txt", "r", stdin);
    // freopen("out.txt", "w", stdout);
    cin.tie(0)->sync_with_stdio(0);
    cout.tie(0)->sync_with_stdio(0);
    int T = 1;
    // cin >> T;
    precompute();
    while (T--) {
        solve();
    }
    cerr << "It's ok Long, Good for now !!" << nln;
    return 0;
}

File: Competitive Programming/codeforces/training/D_General_Weighted_Max_Matching.py
# Copyright (c) 2023, Le Duc Phuc Long

# If you don't think twice, you have to code twice.

# Import session
import sys
#input = sys.stdin.readline
from collections import defaultdict
from math import *

############ ---- Input Functions ---- ############
def inp():
    return int(input())

def inlt():
    return list(map(int, input().split()))

def instr():
    return list(input().strip())

def invr():
    return map(int, input().split())

############ ---- Other Functions ---- ############
# Precompute

# IO
#input = sys.stdin.readline
# sys.stdin = open('in.txt', 'r')
# sys.stdout = open('out.txt', 'w')
# Main function
n = inp()
drr = [[0 for x in range(n)] for y in range(n)]
for i in range(n-1):
    d = inlt()
    for j in range(i+1, n):
        drr[i][j] = drr[j][i] = d[j-i-1]

def dfs(used):
    if (all(used)):
        return 0
    v = used.index(False)
    used[v] = True
    total = 0
    for i in range(n):
        if not used[i]:
            used[i] = True
            total = max(total, dfs(used)+drr[v][i])
            used[i] = False
    used[v] = False
    return total

ans = 0
used = [False]*(n)

if n % 2 == 0:
    ans = dfs(used)
else:
    for i in range(n):
        used[i] = True
        ans = max(ans, dfs(used))
        used[i] = False
    
print(ans)
